File: day20/python/particles.py
#!/usr/bin/python3

import sys
from math import sqrt
from operator import add

class Particle:
    def __init__(self, position, velocity, acceleration):
        self.position = position
        self.velocity = velocity
        self.acceleration = acceleration
        self.destroyed = False

    def update(self):
        self.velocity = list(map(add, self.velocity, self.acceleration))
        self.position = list(map(add, self.position, self.velocity))

    def calculateAccelerationMagnitude(self):
        return sqrt(sum([a ** 2 for a in self.acceleration]))

    def markDestroyedIfCollides(self, particle):
        if self.position == particle.position:
            self.destroyed = particle.destroyed = True

if len(sys.argv) != 2:
    print('Provide input file as command line argument!')
else:
    particles = []
    for particle in open(sys.argv[1]).read().splitlines():
        p_begin = particle.index('<')
        p_end = particle.index('>', p_begin)
        v_begin = particle.index('<', p_end)
        v_end = particle.index('>', v_begin)
        a_begin = particle.index('<', v_end)
        a_end = particle.index('>', a_begin)
        p = [int(i) for i in particle[p_begin + 1:p_end].split(',')]
        v = [int(i) for i in particle[v_begin + 1:v_end].split(',')]
        a = [int(i) for i in particle[a_begin + 1:a_end].split(',')]
        particles.append(Particle(p, v, a))

    numberOfParticles = len(particles)

    # PART 1: works only when there is only one particle with lowest value of acceleration magnitude
    magnitudes = [particle.calculateAccelerationMagnitude() for particle in particles]
    print('Index of particle closest to (0,0,0) in the long term:', magnitudes.index(min(magnitudes)))

    # PART 2: dirty solution, simulate as long as number of particles isn't changing anymore
    for _ in range(10000):
        for i in range(numberOfParticles):
            for j in range(i + 1, numberOfParticles):
                particles[i].markDestroyedIfCollides(particles[j])
        particles = [particle for particle in particles if not particle.destroyed]

        for i, particle in enumerate(particles):
            particle.update()

        numberOfParticles = len(particles)
        print(numberOfParticles)

File: day20/rust/particles.rs
use std::fs;

struct Particle {
    position: (i32, i32, i32),
    velocity: (i32, i32, i32),
    acceleration: (i32, i32, i32),
    destroyed: bool,
}

impl Particle {
    fn update(&mut self) {
        self.velocity.0 += self.acceleration.0;
        self.velocity.1 += self.acceleration.1;
        self.velocity.2 += self.acceleration.2;
        self.position.0 += self.velocity.0;
        self.position.1 += self.velocity.1;
        self.position.2 += self.velocity.2;
    }

    fn calculate_acceleration_magnitude(&self) -> f64 {
        ((self.acceleration.0.pow(2) + self.acceleration.1.pow(2) + self.acceleration.2.pow(2))
            as f64)
            .sqrt()
    }

    fn mark_destroyed_if_collide(&mut self, x: i32, y: i32, z: i32) {
        if self.position.0 == x && self.position.1 == y && self.position.2 == z {
            self.destroyed = true;
        }
    }
}

fn main() {
    let content = fs::read_to_string("../input").expect("file not found");
    let mut particles = Vec::new();

    let mut magnitudes = Vec::new();
    for line in content.lines() {
        let p_begin = line.find('<').unwrap();
        let p_end = line.find('>').unwrap();
        let position = &line[p_begin + 1..p_end];
        let mut position = position.split(',');

        let v_begin = &line[p_end + 1..].find('<').unwrap() + p_end + 1;
        let v_end = &line[p_end + 1..].find('>').unwrap() + p_end + 1;
        let velocity = &line[v_begin + 1..v_end];
        let mut velocity = velocity.split(',');

        let a_begin = &line[v_end + 1..].find('<').unwrap() + v_end + 1;
        let a_end = &line[v_end + 1..].find('>').unwrap() + v_end + 1;
        let acceleration = &line[a_begin + 1..a_end];
        let mut acceleration = acceleration.split(',');

        particles.push(Particle {
            position: (
                position.next().unwrap().parse().unwrap(),
                position.next().unwrap().parse().unwrap(),
                position.next().unwrap().parse().unwrap(),
            ),
            velocity: (
                velocity.next().unwrap().parse().unwrap(),
                velocity.next().unwrap().parse().unwrap(),
                velocity.next().unwrap().parse().unwrap(),
            ),
            acceleration: (
                acceleration.next().unwrap().parse().unwrap(),
                acceleration.next().unwrap().parse().unwrap(),
                acceleration.next().unwrap().parse().unwrap(),
            ),
            destroyed: false,
        });
        magnitudes.push(particles.last().unwrap().calculate_acceleration_magnitude());
    }

    let min_element = magnitudes
        .iter()
        .min_by(|x, y| {
            if x < y {
                return std::cmp::Ordering::Less;
            }
            return std::cmp::Ordering::Greater;
        }).unwrap();
    let min_index = magnitudes
        .iter()
        .position(|element| *element == *min_element)
        .unwrap();
    println!("{}", min_index);

    let mut number_of_particles = particles.len();
    for _ in 0..1_000 {
        for i in 0..number_of_particles {
            for j in 0..number_of_particles {
                if i != j {
                    let (x, y, z) = (
                        particles[j].position.0,
                        particles[j].position.1,
                        particles[j].position.2,
                    );
                    particles[i].mark_destroyed_if_collide(x, y, z);
                }
            }
        }

        particles = particles
            .into_iter()
            .filter(|ref i| !i.destroyed)
            .collect::<Vec<_>>();
        particles.iter_mut().for_each(|particle| particle.update());
        number_of_particles = particles.len();
        println!("{}", number_of_particles);
    }
}
